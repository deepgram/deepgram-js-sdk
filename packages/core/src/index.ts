import { DeepgramClientOptions, IKeyFactory } from "./lib/types/DeepgramClientOptions";
import { DeepgramVersionError } from "./lib/errors";
import DeepgramClient from "./DeepgramClient";

/**
 * This class is deprecated and should not be used. It throws a `DeepgramVersionError` when instantiated.
 *
 * @deprecated
 * @see https://dpgr.am/js-v3
 */
class Deepgram {
  constructor(protected apiKey: string, protected apiUrl?: string, protected requireSSL?: boolean) {
    throw new DeepgramVersionError();
  }
}

/**
 * Creates a new Deepgram client instance.
 *
 * @param {DeepgramClientArgs} args - Arguments to pass to the Deepgram client constructor.
 * @returns A new Deepgram client instance.
 */
function createClient(): DeepgramClient;
function createClient(key?: string | IKeyFactory): DeepgramClient;
function createClient(options?: DeepgramClientOptions): DeepgramClient;
function createClient(key?: string | IKeyFactory, options?: DeepgramClientOptions): DeepgramClient;
function createClient(
  keyOrOptions?: string | IKeyFactory | DeepgramClientOptions,
  options?: DeepgramClientOptions
): DeepgramClient {
  let resolvedOptions: DeepgramClientOptions = {};

  if (typeof keyOrOptions === "string" || typeof keyOrOptions === "function") {
    if (typeof options === "object") {
      resolvedOptions = options;
    }

    resolvedOptions.key = keyOrOptions;
  } else if (typeof keyOrOptions === "object") {
    resolvedOptions = keyOrOptions;
  }

  return new DeepgramClient(resolvedOptions);
}

export { createClient, DeepgramClient, Deepgram };

/**
 * Helpful exports.
 */
export * from "./packages";
export * from "./lib/types";
export * from "./lib/enums";
export * from "./lib/constants";
export * from "./lib/errors";
export * from "./lib/helpers";

/**
 * Captions. These will be tree-shaken if unused.
 *
 * @see https://github.com/deepgram/deepgram-node-captions
 *
 * import/export declarations don't do anything but set up an alias to the
 * exported variable, they do not count as a "use". Given their semantics,
 * they are tracked specially by any bundler and will not adversely affect
 * tree-shaking.
 */
export { webvtt, srt } from "@deepgram/captions";
