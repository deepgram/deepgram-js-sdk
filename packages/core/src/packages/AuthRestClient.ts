import { isDeepgramError } from "../lib/errors";
import type { DeepgramResponse } from "../lib/types/DeepgramResponse";
import type { GrantTokenResponse } from "../lib/types/GrantTokenResponse";
import type { GrantTokenSchema } from "../lib/types/GrantTokenSchema";
import { AbstractRestClient } from "./AbstractRestClient";

export class AuthRestClient extends AbstractRestClient {
  public namespace: string = "auth";

  /**
   * Generates a new temporary token for the Deepgram API.
   * @param options Optional configuration options for the token generation. Includes ttl_seconds to set token expiration.
   * @param endpoint Optional custom endpoint to use for the request. Defaults to ":version/auth/grant".
   * @returns Object containing the result of the request or an error if one occurred. Result will contain access_token and expires_in properties.
   */
  public async grantToken(
    options: GrantTokenSchema = {},
    endpoint = ":version/auth/grant"
  ): Promise<DeepgramResponse<GrantTokenResponse>> {
    try {
      const requestUrl = this.getRequestUrl(endpoint);
      const body = JSON.stringify(options);
      const result: GrantTokenResponse = await this.post(requestUrl, body, {
        headers: { "Content-Type": "application/json" },
      }).then((result) => result.json());

      return { result, error: null };
    } catch (error) {
      if (isDeepgramError(error)) {
        return { result: null, error };
      }

      throw error;
    }
  }
}
