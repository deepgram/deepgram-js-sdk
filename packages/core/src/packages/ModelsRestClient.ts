import { isDeepgramError } from "../lib/errors";
import {
  DeepgramResponse,
  GetModelResponse,
  GetModelsResponse,
  GetModelsSchema,
} from "../lib/types";
import { AbstractRestClient } from "./AbstractRestClient";

/**
 * Represents a REST client for interacting with the Deepgram API.
 *
 * The `ModelsRestClient` class provides methods for interacting with the Deepgram API to retrieve information about available models.
 * @extends AbstractRestClient
 */
export class ModelsRestClient extends AbstractRestClient {
  public namespace: string = "models";

  /**
   * Retrieves a list of all available models.
   *
   * @param endpoint - (optional) The endpoint to request.
   * @returns A promise that resolves with the response from the Deepgram API.
   * @example
   * ```typescript
   * import { createClient } from "@deepgram/sdk";
   *
   * const deepgram = createClient(DEEPGRAM_API_KEY);
   * const { result: models, error } = deepgram.models.getAll();
   *
   * if (error) {
   *   console.error(error);
   * } else {
   *   console.log(models);
   * }
   * ```
   */
  async getAll(
    endpoint = ":version/models",
    options: GetModelsSchema = {}
  ): Promise<DeepgramResponse<GetModelsResponse>> {
    try {
      const requestUrl = this.getRequestUrl(endpoint, {}, options);
      const result: GetModelsResponse = await this.get(requestUrl).then((result) => result.json());

      return { result, error: null };
    } catch (error) {
      if (isDeepgramError(error)) {
        return { result: null, error };
      }

      throw error;
    }
  }

  /**
   * Retrieves information about a specific model.
   *
   * @param modelId - The UUID of the model to retrieve.
   * @param endpoint - (optional) The endpoint to request.
   * @returns A promise that resolves with the response from the Deepgram API.
   * @example
   * ```typescript
   * import { createClient } from "@deepgram/sdk";
   *
   * const deepgram = createClient(DEEPGRAM_API_KEY);
   * const { result: model, error } = deepgram.models.getModel("modelId");
   *
   * if (error) {
   *   console.error(error);
   * } else {
   *   console.log(model);
   * }
   * ```
   */
  async getModel(
    modelId: string,
    endpoint = ":version/models/:modelId"
  ): Promise<DeepgramResponse<GetModelResponse>> {
    try {
      const requestUrl = this.getRequestUrl(endpoint, { modelId });
      const result: GetModelResponse = await this.get(requestUrl).then((result) => result.json());

      return { result, error: null };
    } catch (error) {
      if (isDeepgramError(error)) {
        return { result: null, error };
      }

      throw error;
    }
  }
}
