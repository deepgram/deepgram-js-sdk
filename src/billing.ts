import { BalanceList, Balance, RequestFunction } from "./types";

export class Billing {
  constructor(
    private _credentials: string,
    private _apiUrl: string,
    private _requireSSL: boolean,
    private _request: RequestFunction
  ) {}

  /**
   * Retrieves list of balance info of the specified project.
   * @param {string} projectId Unique identifier of the project
   * @param {string} endpoint Custom API endpoint
   *
   * @returns {Promise<BalanceList>}
   */
  async listBalances(
    projectId: string,
    endpoint = "v1/projects"
  ): Promise<BalanceList> {
    return this._request(
      "GET",
      this._credentials,
      this._apiUrl,
      this._requireSSL,
      `/${endpoint}/${projectId}/balances`
    );
  }

  /**
   * Retrieves balance info of a specified balance_id in the specified project.
   * @param {string} projectId Unique identifier of the project
   * @param {string} balanceId Unique identifier of the balance
   * @param {string} endpoint Custom API endpoint
   *
   * @returns {Promise<Balance>}
   */
  async getBalance(
    projectId: string,
    balanceId: string,
    endpoint = "v1/projects"
  ): Promise<Balance> {
    return this._request(
      "GET",
      this._credentials,
      this._apiUrl,
      this._requireSSL,
      `/${endpoint}/${projectId}/balances/${balanceId}`
    );
  }
}
