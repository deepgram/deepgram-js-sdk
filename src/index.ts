import DeepgramClient from "./DeepgramClient";
import { DeepgramVersionError } from "./lib/errors";
import type { DeepgramClientOptions } from "./lib/types";

/**
 * Major version fallback error
 */
class Deepgram {
  constructor(protected apiKey: string, protected apiUrl?: string, protected requireSSL?: boolean) {
    throw new DeepgramVersionError();
  }
}

/**
 * Creates a new Deepgram Client.
 */
const createClient = (apiKey: string, options: DeepgramClientOptions = {}): DeepgramClient => {
  return new DeepgramClient(apiKey, options);
};

export { createClient, DeepgramClient, Deepgram };

/**
 * Helpful exports.
 */
export * from "./packages";
export * from "./lib/types";
export * from "./lib/enums";
export * from "./lib/constants";
export * from "./lib/errors";
export * from "./lib/helpers";

/**
 * Captions. These will be tree-shaken if unused.
 *
 * @see https://github.com/deepgram/deepgram-node-captions
 *
 * import/export declarations don't do anything but set up an alias to the
 * exported variable, they do not count as a "use". Given their semantics,
 * they are tracked specially by any bundler and will not adversely affect
 * tree-shaking.
 */
export { webvtt, srt } from "@deepgram/captions";
