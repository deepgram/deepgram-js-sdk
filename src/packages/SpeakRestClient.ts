import { DeepgramError, DeepgramUnknownError } from "../lib/errors";
import { isTextSource } from "../lib/helpers";
import { SpeakSchema, TextSource } from "../lib/types";
import { AbstractRestClient } from "./AbstractRestClient";

/**
 * Provides a client for interacting with the Deepgram Text-to-Speech API.
 */
export class SpeakRestClient extends AbstractRestClient {
  public namespace: string = "speak";
  public result: undefined | Response;

  /**
   * Sends a request to the Deepgram Text-to-Speech API to generate audio from the provided text source.
   *
   * @param source - The text source to be converted to audio.
   * @param options - Optional configuration options for the text-to-speech request.
   * @param endpoint - The API endpoint to use for the request. Defaults to ":version/speak".
   * @returns A promise that resolves to the SpeakRestClient instance, which can be used to retrieve the response headers and body.
   * @throws {DeepgramError} If the text source type is unknown.
   * @throws {DeepgramUnknownError} If the request was made before a previous request completed.
   * @see https://developers.deepgram.com/reference/text-to-speech-api
   */
  async request(
    source: TextSource,
    options?: SpeakSchema,
    endpoint = ":version/speak"
  ): Promise<SpeakRestClient> {
    let body;

    if (isTextSource(source)) {
      body = JSON.stringify(source);
    } else {
      throw new DeepgramError("Unknown transcription source type");
    }

    const requestUrl = this.getRequestUrl(
      endpoint,
      {},
      { ...{ model: "aura-asteria-en" }, ...options }
    );
    this.result = await this.post(requestUrl, body, {
      headers: { Accept: "audio/*", "Content-Type": "application/json" },
    });

    return this;
  }

  /**
   * Retrieves the response body as a readable stream.
   *
   * @returns A promise that resolves to the response body as a readable stream, or `null` if no request has been made yet.
   * @throws {DeepgramUnknownError} If a request has not been made yet.
   */
  async getStream(): Promise<ReadableStream<Uint8Array> | null> {
    if (!this.result)
      throw new DeepgramUnknownError("Tried to get stream before making request", "");

    return this.result.body;
  }

  /**
   * Retrieves the response headers from the previous request.
   *
   * @returns A promise that resolves to the response headers, or throws a `DeepgramUnknownError` if no request has been made yet.
   */
  async getHeaders(): Promise<Headers> {
    if (!this.result)
      throw new DeepgramUnknownError("Tried to get headers before making request", "");

    return this.result.headers;
  }
}
