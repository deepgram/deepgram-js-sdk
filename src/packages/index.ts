export { AbstractClient } from "./AbstractClient";
export { AbstractRestfulClient } from "./AbstractRestfulClient";
export { AbstractWsClient } from "./AbstractWsClient";
export { ListenClient } from "./ListenClient";
export { LiveClient } from "./LiveClient";
export { ManageClient } from "./ManageClient";
export { OnPremClient } from "./OnPremClient";
export { PrerecordedClient } from "./PrerecordedClient";
