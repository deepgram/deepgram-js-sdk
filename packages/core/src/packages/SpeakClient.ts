import { AbstractClient } from "./AbstractClient";
import { SpeakLiveClient } from "./SpeakLiveClient";
import { SpeakRestClient } from "./SpeakRestClient";
import { SpeakSchema } from "../lib/types";
import { TextSource } from "../lib/types";

/**
 * The `SpeakClient` class extends the `AbstractClient` class and provides access to the "speak" namespace.
 * It exposes two methods:
 *
 * 1. `request()`: Returns a `SpeakRestClient` instance for interacting with the rest speak API.
 * 2. `live(ttsOptions: SpeakSchema = {}, endpoint = ":version/speak")`: Returns a `SpeakLiveClient` instance for interacting with the live speak API, with the provided TTS options and endpoint.
 */
export class SpeakClient extends AbstractClient {
  public namespace: string = "speak";

  /**
   * Returns a `SpeakRestClient` instance for interacting with the rest speak API.
   */
  public request(source: TextSource, options?: SpeakSchema, endpoint = ":version/speak") {
    const client = new SpeakRestClient(this.options);

    return client.request(source, options, endpoint);
  }

  /**
   * Returns a `SpeakLiveClient` instance for interacting with the live speak API, with the provided TTS options and endpoint.
   * @param {SpeakSchema} [ttsOptions={}] - The TTS options to use for the live speak API.
   * @param {string} [endpoint=":version/speak"] - The endpoint to use for the live speak API.
   * @returns {SpeakLiveClient} - A `SpeakLiveClient` instance for interacting with the live speak API.
   */
  public live(ttsOptions: SpeakSchema = {}, endpoint: string = ":version/speak"): SpeakLiveClient {
    return new SpeakLiveClient(this.options, ttsOptions, endpoint);
  }
}
