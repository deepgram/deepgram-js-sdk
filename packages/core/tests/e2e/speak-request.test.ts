import { createClient } from "../../src/index";
import { SpeakRestClient } from "../../src/packages/SpeakRestClient";
import { structureOnlySerializer, setupApiMocks, cleanupApiMocks } from "../__utils__";
import { testTextSources, commonTTSOptions } from "../__fixtures__/speak";

describe("speak request E2E", () => {
  let deepgram: ReturnType<typeof createClient>;

  beforeAll(() => {
    // Set up API mocks (only active when not updating snapshots)
    setupApiMocks();

    // Use mock API key when mocking, real one when updating snapshots
    const apiKey = process.env.DEEPGRAM_API_KEY || "mock-api-key";
    deepgram = createClient(apiKey);

    // Add our custom serializer
    expect.addSnapshotSerializer(structureOnlySerializer);
  });

  afterAll(() => {
    // Clean up mocks
    cleanupApiMocks();
  });

  it("should generate speech from text and return audio data", async () => {
    const speakClient = await deepgram.speak.request(testTextSources.greeting, commonTTSOptions);

    // Verify we got a response
    expect(speakClient).toBeTruthy();
    expect(speakClient.result).toBeTruthy();

    // Type guard to ensure result is not null for subsequent operations
    if (!speakClient.result) {
      throw new Error("Result should not be null after toBeTruthy check");
    }

    // Verify the response structure
    expect(speakClient.result.ok).toBe(true);
    expect(speakClient.result.status).toBe(200);

    // Verify we can get the audio stream
    const audioStream = await speakClient.getStream();
    expect(audioStream).toBeTruthy();

    // Verify headers contain expected TTS metadata
    const headers = await speakClient.getHeaders();
    expect(headers).toBeTruthy();
    expect(headers.get("content-type")).toContain("audio/");

    // Check for Deepgram-specific headers (if present)
    const contentLength = headers.get("content-length");
    if (contentLength) {
      expect(parseInt(contentLength)).toBeGreaterThan(0);
    }

    // Test the structure with snapshot
    const headerObj = Object.fromEntries(headers.entries());
    expect(headerObj).toMatchSnapshot("speak-request-response-headers");

    // Verify audio stream is readable
    expect(audioStream).toBeInstanceOf(ReadableStream);
  }, 30000); // 30 second timeout for API call

  it("should handle different TTS options", async () => {
    const speakClient = await deepgram.speak.request(testTextSources.longText, {
      model: "aura-2-thalia-en",
      encoding: "linear16",
      sample_rate: 48000,
    });

    // Verify we got a response with the configured options
    expect(speakClient).toBeTruthy();
    expect(speakClient.result).toBeTruthy();

    if (!speakClient.result) {
      throw new Error("Result should not be null after toBeTruthy check");
    }

    expect(speakClient.result.ok).toBe(true);
    expect(speakClient.result.status).toBe(200);

    // Verify audio stream is available
    const audioStream = await speakClient.getStream();
    expect(audioStream).toBeTruthy();
    expect(audioStream).toBeInstanceOf(ReadableStream);

    // Verify headers
    const headers = await speakClient.getHeaders();
    expect(headers).toBeTruthy();
    expect(headers.get("content-type")).toBeTruthy();
  }, 30000);

  it("should handle multiline text input", async () => {
    const speakClient = await deepgram.speak.request(testTextSources.multiline, commonTTSOptions);

    // Verify we got a response
    expect(speakClient).toBeTruthy();
    expect(speakClient.result).toBeTruthy();

    if (!speakClient.result) {
      throw new Error("Result should not be null after toBeTruthy check");
    }

    expect(speakClient.result.ok).toBe(true);

    // Verify we can access the audio data
    const audioStream = await speakClient.getStream();
    expect(audioStream).toBeTruthy();

    const headers = await speakClient.getHeaders();
    expect(headers).toBeTruthy();
    expect(headers.get("content-type")).toContain("audio/");
  }, 30000);

  it("should reject empty text source", async () => {
    // Empty text should be rejected by the client validation
    await expect(deepgram.speak.request({ text: "" }, commonTTSOptions)).rejects.toThrow(
      "Unknown transcription source type"
    );
  }, 30000);

  it("should handle various audio encodings", async () => {
    const encodingTests = [
      { encoding: "mp3" as const },
      { encoding: "linear16" as const, sample_rate: 48000 },
      { encoding: "opus" as const },
    ];

    for (const options of encodingTests) {
      const speakClient = await deepgram.speak.request(
        { text: "Testing encoding: " + options.encoding },
        {
          model: "aura-2-thalia-en",
          ...options,
        }
      );

      expect(speakClient).toBeTruthy();
      expect(speakClient.result).toBeTruthy();

      if (!speakClient.result) {
        throw new Error("Result should not be null after toBeTruthy check");
      }

      expect(speakClient.result.ok).toBe(true);

      const audioStream = await speakClient.getStream();
      expect(audioStream).toBeTruthy();
    }
  }, 30000);

  it("should preserve special characters and punctuation", async () => {
    const specialTextSource = {
      text: "Hello! How are you? I'm fine. Testing 123, $50.00, and Ã©mojis ðŸŽµ",
    };

    const speakClient = await deepgram.speak.request(specialTextSource, commonTTSOptions);

    expect(speakClient).toBeTruthy();
    expect(speakClient.result).toBeTruthy();

    if (!speakClient.result) {
      throw new Error("Result should not be null after toBeTruthy check");
    }

    expect(speakClient.result.ok).toBe(true);

    const audioStream = await speakClient.getStream();
    expect(audioStream).toBeTruthy();

    const headers = await speakClient.getHeaders();
    expect(headers).toBeTruthy();
    expect(headers.get("content-type")).toContain("audio/");
  }, 30000);

  it("should throw error when trying to getStream before making a request", async () => {
    const apiKey = process.env.DEEPGRAM_API_KEY || "mock-api-key";
    const speakClient = new SpeakRestClient({ key: apiKey });

    await expect(speakClient.getStream()).rejects.toThrow(
      "Tried to get stream before making request"
    );
  });

  it("should throw error when trying to getHeaders before making a request", async () => {
    const apiKey = process.env.DEEPGRAM_API_KEY || "mock-api-key";
    const speakClient = new SpeakRestClient({ key: apiKey });

    await expect(speakClient.getHeaders()).rejects.toThrow(
      "Tried to get headers before making request"
    );
  });
});
