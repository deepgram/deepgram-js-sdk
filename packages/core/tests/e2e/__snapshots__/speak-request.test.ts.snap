// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`speak request E2E should generate speech from text and return audio data: speak-request-response-headers 1`] = `
{
  "content-length": "<string>",
  "content-type": "<string>",
  "x-dg-model": "<string>",
  "x-dg-request-id": "<string>"
}
`;
